processing/src/main/java/org/apache/druid/math/expr/ConstantExpr.java

```diff
@@ -100,6 +106,67 @@ public String stringify()
   {
     return toString();
   }
+
+  @Override
+  public final ExprEval eval(ObjectBinding bindings)
+  {
+    return realEval();
+  }
+
+  protected abstract ExprEval<T> realEval();
+
+
+  @Override
+  public Expr toSingleThreaded()
+  {
+    return new ExprEvalBasedConstantExpr<T>(realEval());
+  }
+
+  /**
+   * Constant expression based on a concreate ExprEval.
+   *
+   * Not multi-thread safe.
+   */
+  @NotThreadSafe
+  @SuppressWarnings("Immutable")
+  private static final class ExprEvalBasedConstantExpr<T> extends ConstantExpr<T>
+  {
+    private final ExprEval<T> eval;
+
+    private ExprEvalBasedConstantExpr(ExprEval<T> eval)
+    {
+      super(eval.type(), eval.value);
+      this.eval = eval;
+    }
+
+    @Override
+    public ExprEval<T> realEval()
```

Comment:
Could stay `protected`? I got briefly confused as to why this was `public` but I think it doesn't need to be. Similar comments for the others.

Reply:
of course - my intention was to have them protected...